x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/TransportTermEnumAction.java

```diff
@@ -0,0 +1,605 @@
+            terms.add(term.getTerm());
+            if (terms.size() == size) {
+                break;
+            }
+        }
+        return new TermEnumResponse(terms, nodesResponses.length(), successfulShards, failedShards, shardFailures, complete);
+    }
+
+    static class TermCountPriorityQueue extends PriorityQueue<TermCount> {
+
+        TermCountPriorityQueue(int maxSize) {
+            super(maxSize);
+        }
+
+        @Override
+        protected boolean lessThan(TermCount a, TermCount b) {
+            return a.getDocCount() < b.getDocCount();
+        }
+
+    }
+    
+    protected NodeTermEnumResponse dataNodeOperation(NodeTermEnumRequest request, Task task) throws IOException {
+        List<TermCount> termsList = new ArrayList<>();
+        String error = null;
+        
+        // DLS/FLS check copied from ResizeRequestInterceptor
+        // MH code - not sure this is the right context
+        ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
+        final XPackLicenseState frozenLicenseState = licenseState.copyCurrentLicenseState();
+
+        long timeout_millis = request.timeout();
+        long scheduledEnd = request.shardStartedTimeMillis() + timeout_millis;
+
+        ArrayList<IndexReader> shardReaders = new ArrayList<>();
+        ArrayList<Closeable> openedResources = new ArrayList<>();
+        try {
+            for (ShardId shardId : request.shardIds()) {
+                // Check we haven't just arrived on a node and time is up already.
+                if (System.currentTimeMillis() > scheduledEnd) {
+                    return new NodeTermEnumResponse(request.nodeId(), termsList, error, false);
+                }
+                final IndexService indexService = indicesService.indexServiceSafe(shardId.getIndex());
+                final IndexShard indexShard = indexService.getShard(shardId.getId());
+    
+                Engine.Searcher searcher = indexShard.acquireSearcher(Engine.CAN_MATCH_SEARCH_SOURCE);
+                openedResources.add(searcher);
+                    final SearchExecutionContext queryShardContext = indexService.newSearchExecutionContext(
+                        shardId.id(),
+                        0,
+                        searcher,
+                        request::shardStartedTimeMillis,
+                        null,
+                        Collections.emptyMap()
+                    );
+                    if (canAccess(shardId.getIndexName(), request.field(), frozenLicenseState, threadContext) && 
+                        canMatchShard(shardId, request, queryShardContext)) {
+                        shardReaders.add(searcher.getTopReaderContext().reader());
+                    }
+                
+            }
+            MultiReader multiReader = new MultiReader(shardReaders.toArray(new IndexReader[0]), false);
+
+            Terms terms = MultiTerms.getTerms(multiReader, request.field());
```

Comment:
I realize this is WIP, but had one high-level thought that might affect code structure/ naming. It'd be great to encapsulate this term loading logic into `MappedFieldType` instead of interacting with Lucene data structures directly here. We have been trying to do this consistently, and now delegate query creation, doc value loading, and other data fetching tasks to `MappedFieldType`. Some benefits:
* Field types can decide to handle the call differently. For example `constant_keyword` could support this functionality (even though it doesn't write them in the index).
* Field aliases will be handled correctly, since we resolve them when looking up `MappedFieldType`
* If there's any change to logic because of an index version change, we can consolidate that in one place

Reply:
I did suggest that MappedFieldType delegation to Jim the other day but he was not so keen but I don't recall exactly why.
Possibly deferring the change to a later development phase or a concern over losing the optimisation we get from having a single MultiReader view on multiple shards/indices. Do you remember @jimczi ?

Reply:
+1 to add the logic to return `Terms` or `TermsEnum` in the `MappedFiedlType`. Mark we've discussed this already and I remember saying that it was important in order to handle `constant_keyword` and `flattened` correctly ;).
It could return an empty or null terms by default to make it clear that the API is not mandatory for each field type.

Reply:
What's your feeling on the timer question in my (edited) comment above?

Reply:
I don't understand the relation with completeness ? We're talking of a terms enum that can only be used to search by prefix so the proposal is something like:
`TermsEnum getTerms(String prefix, boolean caseSensitive)` or `Terms getTerms()` where we'd leave more room for the caller. So there's no need for each field type to implement anything regarding early termination or timeout. It's the terms enum API that wraps all these `TermsEnum` in a `MultiTermsEnum` and knows where to stop and if the result is complete or not.

Reply:
It's allowing for anything that might be expensive e.g. runtime fields or wildcard fields that choose to attempt _something_ within the provided time constraints (timeout being something you said should be a parameter in this API).
They'd have to gather a sample of docs _before_ returning a TermsEnum which enumerates over the discovered terms.

Reply:
I think runtime fields should just return an empty TermsEnum here? Probably the same with wildcard fields tbh, they don't have a sensible terms dictionary to respond from.  Or wildcard fields could return a MultiTerms over the term dictionaries for all their doc value leaf dictionaries, which will be fast to construct even if its slower to iterate over

Reply:
I don't think wildcard could do anything _trustworthy_ whether it chooses to block on the construction or the logic that determines what to return first from terms.next();
That's kind of the point here - we want to communicate to users if there's the possibility of any false negatives in the results being presented. A false negative is bad because we don't want users to walk away without searching because the autocomplete results have (falsely) suggested to them there's nothing to be found.  
We have several policy options:
1) Kibana teaches users to distrust _all_ suggestions
2) Users generally trust results but Kibana flags any incomplete suggestions. Field types inform incompleteness by either: 
  a) returning all-or-nothing. A null response means users are told there's stuff missing. A non-null response means the set is assumed to be complete
  b) fields types return a subset of  terms and somehow communicates the completeness of this set.

If we assume completeness is useful (and that was a design goal) we need to pick either 2a or 2b for this API contract

Reply:
> It's allowing for anything that might be expensive e.g. runtime fields or wildcard fields that choose to attempt something within the provided time constraints (timeout being something you said should be a parameter in this API).
They'd have to gather a sample of docs before returning a TermsEnum which enumerates over the discovered terms.

Please no, we discussed that already and we said that it's not something we want to provide. If you have a wildcard or a runtime field then you're not eligible to the terms enum API and we return fields. That's the part that should be clear in the javadocs of the method, return the terms enum only if you can provide the feature efficiently.

Reply:
OK. What confused me was the idea that timeout should be a parameter passed to this field type API and how that would be interpreted/used

Reply:
The MappedFieldType interface may need some more thought.
For efficiency's sake the current PR uses a MultiReader on a node to provide a single view across potentially many shards. While quick, there are several issues here:

1) **Integer doc count limit** - if we ever want to sort by popularity, the TermsEnum.docFreq() returns an integer which will potentially overflow if we try use a single view across multiple shards. Using this Lucene-based abstraction over multiple Lucene indices is not going to scale (TODO - fix in Lucene or copy Lucene's MultiTermsEnum ?)
2) **Resource management** the current one-size-fits-all logic  has a try-finally block to close readers used while iterating across Terms - the proposed new MappedFieldType API above has not considered how to close any underlying resources accessed by the returned Terms object.
3) **Some, all or no fusion delegation?** - an open question is if the underlying field types are expected to perform any fusion of data across shards in the view offered by the returned TermsEnum. How do they know which Lucene  index or indices they should access? Are they passed one shardID or multiple? How would the caller determine that 2 field types e.g. the `keyword` field for `host` in index `logs08-02-2021` and the same field in index `logs09-02-2021` were merge-able by one of the field types? We would, of course, assume that MappedFieldTypes of different classes would have different implementations and so their TermsEnums would have to be merged by the caller doing the collection on a node. How would the caller fuse multiple TermEnums? The Lucene MultiTermsEnum class is constructed using `ReaderSlice` objects.

Reply:
Discussed with Jim. We'll go with

    TermsEnum getTerms(String prefix, boolean caseSensitive, SearchExecutionContext context)

The caller calls this on the relevant field type for every shard and merges the TermsEnum objects using a new MultiTermsEnum class that differs from the Lucene one in that it:
1) Takes a number of TermsEnum objects rather than ReaderSlice objects
2) Uses longs not ints for summing doc frequencies
3) Doesn't implement TermsEnum, so doesn't offer all its methods like ord, postings etc Instead it has next(), term() and docFreq() only